docs/proposals/vip-3.md

```diff
@@ -0,0 +1,226 @@
+   too expensive, or by reserving some allotment periodically, which is complicated).
+2. Let the Rust server and Java server each have the full quota amount and not need any interaction between each other 
+   for sharing it. This is much simpler to implement, but the downside is that if clients were to perfectly balance
+   their traffic across the two ports, they could get up to twice their configured quota.
+
+Since the migration is expected to be controlled by the operator, and should only be activated for users of the Fast
+Client (i.e. not for users that run a mix of Thin Clients and Fast Clients), the edge case solved by the more 
+complicated solution 1 does not seem likely to come up. Therefore, the proposal is to pick the simpler solution 2.
+
+Also, it can be considered whether quota should be part of the MVP at all. If the Rust server mode will be tried in 
+dedicated clusters at first, then quota may not even be a must-have (though of course it will be needed eventually).
+
+### Cluster Discovery
+
+The general idea is that when the Rust mode is activated in a server, that server would announce that it has a gRPC 
+endpoint at the port which the Rust server listens to. The announcement can be done by the Java server, to minimize
+complexity on the Rust side. Beyond that, there are many variants on how to make things work end-to-end.
+
+A few alternatives:
+
+1. The Java and Rust servers are all announced to the same resource name. This means there is some gRPC resource which
+   a Fast Client can hit, and that could be served either by Java or Rust, and possibly both. Changing which path the
+   clients go through would require turning on or off the announcement on the Java and/or Rust side, which could be
+   done either via reconfiguring and bouncing the server, or via some admin tooling.
+2. The Java and Rust servers are announced to different resource names. When cluster discovery happens, the client asks
+   where to connect, the backend could decide to return one or the other of these resources, thus directing the traffic
+   fully (at least for that one client instance) into one path or the other. Changing which path the clients go through
+   would require updating the mapping returned by cluster discovery, which would incur the same delay as cluster 
+   migration.
+3. Instead of announcing per-cluster resources, we could announce per-store resources, and then have some admin tooling
+   to enable and disable the announcement for Rust and Java servers on a per-store basis. This could give the most
+   flexibility, since the client would not remain stuck with whatever resource name was returned by the cluster 
+   discovery call done at startup time (and periodically thereafter), but rather could adapt rapidly.
+
+These options provide different levels of granularity of control and speed of change, with finer granularity and faster
+change carrying greater complexity cost.
+
+## Development Milestones
```

Comment:
Venice Server is also using a lot of dependencies, such as Helix, ZK, ACL (maybe some others I don't remember), and I am not sure whether we have rust libraries for all these dependencies.

One strategy I was thinking is to use a push model from the Java process to Rust Server.
It means, Java Process/Server will have a thread to explicitly detect/listen to the changes in Helix/ZK/ACL and push these information in Venice controlled format asynchronously (ASAP) to the rust server, so that Venice could develop these functionalities even Rust Helix lib is not there yet. Once the native Rust support for these dependencies are available, we can switch to them.

Reply:
Regarding ACLs, there is a similar discussion going on here: https://github.com/linkedin/venice/pull/604#discussion_r1303656727

Regarding Helix, I am thinking that ideally the Rust server doesn't need to know about it at all, but maybe that is not quite possible. When the Java side does the JNI call to create/open a RocksDV store-version-partition, the Rust wrapper can remember the fact that this partition exists, and therefore serve it when requested. Likewise when the JNI call comes in to drop a RocksDB partition, the Rust wrapper needs to respond with an error if it receives a query for that. These RocksDB calls would be side-effects of the Java-side Helix state transitions. If that is sufficient, then no extra exchange protocol is needed as far as Helix is concerned...

Where I think we will definitely need to explore the idea you're bringing up is for Read Compute, where we'll need the Rust side to know about schemas, and those can be added over time.

There might be other situations too, such as quota changes, where Java needs to pass Venice-specific info to Rust.